### PR TITLE
dynamic jobProvider and suggestionComposer registration

### PR DESCRIPTION
**Pull request by sperlingxx**
Add registry for jobProvider and suggestionComposer,  so it is easy to add  new `Provider` or `Composer` via init function without modifying existing files. 
In addition, `Provider` includes an additonal method `MutateJob`, which allows some custom mutation on jobSpec before its creation.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/katib/1069)

### CODE REVIEW

File: pkg/controller.v1alpha3/trial/trial_controller_test.go [hunk 1]
```diff
@@ -2,6 +2,7 @@ package trial
 
 import (
 	"bytes"
+	"github.com/kubeflow/katib/pkg/job/v1alpha3/kubeflow"
 	"testing"
```
Comment (gaocegege) [inline pkg/controller.v1alpha3/trial/trial_controller_test.go]: Move to the next group, please.

File: pkg/job/v1alpha3/job/job.go [hunk 2]
```diff
@@ -1,11 +1,16 @@
...
+	"github.com/kubeflow/katib/pkg/apis/controller/trials/v1alpha3"
+	"github.com/kubeflow/katib/pkg/controller.v1alpha3/consts"
+	job "github.com/kubeflow/katib/pkg/job/v1alpha3"
 	commonv1 "github.com/kubeflow/tf-operator/pkg/apis/common/v1"
 	batchv1 "k8s.io/api/batch/v1"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"reflect"
```
Comment (gaocegege) [inline pkg/job/v1alpha3/job/job.go]: Move to the first group, please.

File: pkg/job/v1alpha3/kubeflow/kubeflow.go [hunk 3]
```diff
@@ -1,13 +1,16 @@
...
+	job "github.com/kubeflow/katib/pkg/job/v1alpha3"
 	pytorchv1 "github.com/kubeflow/pytorch-operator/pkg/apis/pytorch/v1"
 	commonv1 "github.com/kubeflow/tf-operator/pkg/apis/common/v1"
 	tfv1 "github.com/kubeflow/tf-operator/pkg/apis/tensorflow/v1"
-
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"reflect"
```
Comment (gaocegege) [inline pkg/job/v1alpha3/kubeflow/kubeflow.go]: Same here.

File: pkg/job/v1alpha3/provider.go [hunk 4]
```diff
@@ -2,36 +2,39 @@ package v1alpha3
...
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 
-	"github.com/kubeflow/katib/pkg/controller.v1alpha3/consts"
-	"github.com/kubeflow/katib/pkg/job/v1alpha3/job"
-	"github.com/kubeflow/katib/pkg/job/v1alpha3/kubeflow"
 	commonv1 "github.com/kubeflow/tf-operator/pkg/apis/common/v1"
 )
 
+var (
+	ProviderRegistry = make(map[string]reflect.Type)
```
Comment (gaocegege) [inline pkg/job/v1alpha3/provider.go]: Can we use Provider instead of reflect.Type?

Comment (sperlingxx) [reply pkg/job/v1alpha3/provider.go]: Done.

File: pkg/job/v1alpha3/kubeflow/kubeflow.go [hunk 5]
```diff
@@ -75,3 +77,32 @@ func (k Kubeflow) IsTrainingContainer(index int, c corev1.Container) bool {
...
+}
+
+func Register() {
+	job.ProviderRegistry[consts.JobKindTF] = &Kubeflow{}
+	job.SupportedJobList[consts.JobKindTF] = schema.GroupVersionKind{
+		Group:   "kubeflow.org",
+		Version: "v1",
+		Kind:    "TFJob",
+	}
+	job.JobRoleMap[consts.JobKindTF] = []string{"job-role", "tf-job-role"}
```
Comment (andreyvelich) [inline pkg/job/v1alpha3/kubeflow/kubeflow.go]: Maybe define these roles in consts or as variables above. What do you think ?

Comment (sperlingxx) [reply pkg/job/v1alpha3/kubeflow/kubeflow.go]: @andreyvelich I agree with u, and it's done.

**Comment by sperlingxx** [comment]
@andreyvelich  I think this PR is ready.

**Comment by gaocegege** [comment]
/approve

Thanks for the PR!

**Pull request merged**